proxy/handlers.go

```diff
@@ -223,36 +227,62 @@ func (p *Proxy) OAuthCallback(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, stateParameter.RedirectURI, http.StatusFound)
 }
 
+// shouldSkipAuthentication contains conditions for skipping authentication.
+// Conditions should be few in number and have strong justifications.
+func (p *Proxy) shouldSkipAuthentication(r *http.Request) bool {
+	pol, foundPolicy := p.policy(r)
+
+	if isCORSPreflight(r) && foundPolicy && pol.CORSAllowPreflight {
+		log.FromRequest(r).Debug().Msg("proxy: skipping authentication for valid CORS preflight request")
+		return true
+	}
+
+	return false
+}
+
+// isCORSPreflight inspects the request to see if this is a valid CORS preflight request.
+// These checks are not exhaustive, because the proxied server should be verifying it as well.
+//
+// See https://www.html5rocks.com/static/images/cors_server_flowchart.png for more info.
+func isCORSPreflight(r *http.Request) bool {
+	_, ok1 := r.Header["Access-Control-Request-Method"]
```

Comment:
You mentioned in your initial issue that the OPTIONS request requires three headers: `Access-Control-Request-Method`, `Access-Control-Request-Headers`, and the `Origin header`.

Is `Access-Control-Request-Headers` required?

See:
- https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0

Reply:
I believe `Access-Control-Request-Headers` is not required.

See steps 3 and 4 of https://fetch.spec.whatwg.org/#cors-preflight-fetch, which seem to indicate that if the request does not contain any [`CORS-unsafe request-header names`](https://fetch.spec.whatwg.org/#cors-unsafe-request-header-names), then `Access-Control-Request-Headers` may be omitted. 

[This flowchart](https://www.html5rocks.com/static/images/cors_server_flowchart.png) appears to corroborate the above.

Reply:
Great thanks.